File: app/src/main/java/io/github/summerdez/asmrplayer/data/remote/DlsiteClient.java
package io.github.summerdez.asmrplayer.data.remote;

import android.text.TextUtils;
import android.webkit.CookieManager;

import io.github.summerdez.asmrplayer.domain.model.DlsiteDownloadOption;
import io.github.summerdez.asmrplayer.domain.model.DlsiteWork;

import java.io.File;
import java.io.IOException;
import java.util.List;

public final class DlsiteClient {
    public static final String LOGIN_URL = DlsiteRemoteConstants.PLAY_BASE_URL + "/library";

    private final DlsiteHttpClient httpClient = new DlsiteHttpClient();
    private final DlsiteWorkRemote workRemote = new DlsiteWorkRemote(httpClient);
    private final DlsiteCoverRemote coverRemote = new DlsiteCoverRemote(httpClient);
    private final DlsiteContentRemote contentRemote = new DlsiteContentRemote(httpClient);

    public boolean hasLoginCookie() {
        String cookie = CookieManager.getInstance().getCookie(DlsiteRemoteConstants.COOKIE_URL);
        if (TextUtils.isEmpty(cookie)) {
            cookie = CookieManager.getInstance().getCookie(DlsiteRemoteConstants.DL_SITE_COOKIE_URL);
        }
        return !TextUtils.isEmpty(cookie)
                && (cookie.contains("play_session")
                || cookie.contains("loginchecked")
                || cookie.contains("uid_jp")
                || cookie.contains("eridjp"));
    }

    public List<DlsiteWork> fetchPurchasedWorks() throws IOException {
        return workRemote.fetchPurchasedWorks();
    }

    public void downloadTo(DlsiteWork work, File targetFile) throws IOException {
        contentRemote.downloadTo(work, targetFile);
    }

    public File downloadCover(DlsiteWork work, File targetDir) throws IOException {
        return coverRemote.downloadCover(work, targetDir);
    }

    public List<DlsiteDownloadOption> fetchDownloadOptions(DlsiteWork work) throws IOException {
        return contentRemote.fetchDownloadOptions(work);
    }

    public List<File> downloadWorkFiles(DlsiteWork work, File workDir) throws IOException {
        return contentRemote.downloadWorkFiles(work, workDir, "");
    }

    public List<File> downloadWorkFiles(DlsiteWork work, File workDir, String downloadOptionId) throws IOException {
        return contentRemote.downloadWorkFiles(work, workDir, downloadOptionId);
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/data/remote/DlsiteClient.kt
package io.github.summerdez.asmrplayer.data.remote

import android.text.TextUtils
import android.webkit.CookieManager
import io.github.summerdez.asmrplayer.domain.model.DlsiteDownloadOption
import io.github.summerdez.asmrplayer.domain.model.DlsiteWork
import java.io.File
import java.io.IOException

class DlsiteClient {
    private val httpClient = DlsiteHttpClient()
    private val workRemote = DlsiteWorkRemote(httpClient)
    private val coverRemote = DlsiteCoverRemote(httpClient)
    private val contentRemote = DlsiteContentRemote(httpClient)

    fun hasLoginCookie(): Boolean {
        var cookie: String? = CookieManager.getInstance().getCookie(DlsiteRemoteConstants.COOKIE_URL)
        if (TextUtils.isEmpty(cookie)) {
            cookie = CookieManager.getInstance().getCookie(DlsiteRemoteConstants.DL_SITE_COOKIE_URL)
        }
        return !TextUtils.isEmpty(cookie) &&
            (
                cookie?.contains("play_session") == true ||
                    cookie?.contains("loginchecked") == true ||
                    cookie?.contains("uid_jp") == true ||
                    cookie?.contains("eridjp") == true
                )
    }

    @Throws(IOException::class)
    fun fetchPurchasedWorks(): List<DlsiteWork> {
        return workRemote.fetchPurchasedWorks()
    }

    @Throws(IOException::class)
    fun downloadTo(work: DlsiteWork?, targetFile: File?) {
        contentRemote.downloadTo(work, targetFile)
    }

    @Throws(IOException::class)
    fun downloadCover(work: DlsiteWork?, targetDir: File?): File {
        return coverRemote.downloadCover(work, targetDir)
    }

    @Throws(IOException::class)
    fun fetchDownloadOptions(work: DlsiteWork?): List<DlsiteDownloadOption> {
        return contentRemote.fetchDownloadOptions(work)
    }

    @Throws(IOException::class)
    fun downloadWorkFiles(work: DlsiteWork?, workDir: File?): List<File> {
        return contentRemote.downloadWorkFiles(work, workDir, "")
    }

    @Throws(IOException::class)
    fun downloadWorkFiles(work: DlsiteWork?, workDir: File?, downloadOptionId: String?): List<File> {
        return contentRemote.downloadWorkFiles(work, workDir, downloadOptionId)
    }

    @Throws(IOException::class)
    fun downloadWorkFiles(
        work: DlsiteWork?,
        workDir: File?,
        downloadOptionId: String?,
        progressListener: DlsiteContentProgressListener?,
    ): List<File> {
        return contentRemote.downloadWorkFiles(work, workDir, downloadOptionId, progressListener)
    }

    companion object {
        const val LOGIN_URL: String = DlsiteRemoteConstants.PLAY_BASE_URL + "/library"
    }
}
